File: JAVA/930. Binary Subarrays With Sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int count = 0, sum = 0;
        Map<Integer, Integer> prefixSumCount = new HashMap<>();
        prefixSumCount.put(0, 1);         
        for (int num : nums) {
            sum += num;            
            if (prefixSumCount.containsKey(sum - goal)) {
                count += prefixSumCount.get(sum - goal);
            }            
            prefixSumCount.put(sum, prefixSumCount.getOrDefault(sum, 0) + 1);
        }       
        return count;
    }
}

File: C++/930. Binary Subarrays With Sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int n = nums.size();
        unordered_map<int, int> prefixSumFreq;
        int prefixSum = 0;
        int count = 0;        
        for (int i = 0; i < n; i++) {
            prefixSum += nums[i];
            if (prefixSum == goal) {
                count++;
            }
            if (prefixSumFreq.count(prefixSum - goal)) {
                count += prefixSumFreq[prefixSum - goal];
            }
            prefixSumFreq[prefixSum]++;
        }        
        return count;
    }
};
